### PR TITLE
[WFLY-10520] Transform classes to replace deprecated API method call with changes from Gail Badner + Scott Marlow

### PR DESCRIPTION
**Pull request by scottmarlow**
https://issues.jboss.org/browse/WFLY-10520

I started working on the doc, that will be separate.

We plan to include additional transformations, just waiting for test cases to prove that we need them (as per https://docs.google.com/document/d/1cAVBdeQXQfpTwH0f_GKSahh_gRd7bKI6kII-wxoMglU/edit?ts=5b45a2c4).

The current transformations are:

- Change calls to org.hibernate.BasicQueryContract.getFlushMode(), to instead call BasicQueryContract.getHibernateFlushMode().
- Change calls to org.hibernate.Session.getFlushMode, to instead call Session.getHibernateFlushMode()
- Change references to Enum org.hibernate.FlushMode.NEVER (0), to FlushMode.MANUAL (0).
- Change calls to org.hibernate.Query.getMaxResults() returning Integer, to instead call org.hibernate.Query.getHibernateMaxResults() (returning Integer).
- Change calls to org.hibernate.Query.setMaxResults(int), to instead call org.hibernate.Query.setHibernateMaxResults(int).
- Change calls to org.hibernate.Query.getFirstResult(int) returning Integer, to instead call org.hibernate.Query.getHibernateFirstResult() (returning Integer).
- Change calls to org.hibernate.Query.setFirstResult(int), to instead call org.hibernate.Query.setHibernateFirstResult(int)

### CODE REVIEW

File: jpa/subsystem/src/main/java/org/jboss/as/jpa/Hibernate51CompatibilityTransformer.java [hunk 1]
```diff
@@ -0,0 +1,157 @@
...
+import org.objectweb.asm.ClassReader;
+import org.objectweb.asm.ClassVisitor;
+import org.objectweb.asm.ClassWriter;
+import org.objectweb.asm.MethodVisitor;
+import org.objectweb.asm.Opcodes;
+
+/**
+ * A class file transformer which makes deployment classes written for Hibernate 5.1 be compatible with Hibernate 5.3.
+ * <p>
+ * TODO: remove asm-commmons artifact from WF as we aren't using it.  Do this before creating a pr for this change.
```
Comment (jamezp) [inline jpa/subsystem/src/main/java/org/jboss/as/jpa/Hibernate51CompatibilityTransformer.java]: Should this TODO be removed?

File: jpa/subsystem/src/main/java/org/jboss/as/jpa/processor/JPAClassFileTransformerProcessor.java [hunk 2]
```diff
@@ -41,6 +44,13 @@
  */
 public class JPAClassFileTransformerProcessor implements DeploymentUnitProcessor {
 
+    private static final boolean hibernate51CompatibilityTransformer = Boolean.getBoolean("Hibernate51CompatibilityTransformer");
```
Comment (jamezp) [inline jpa/subsystem/src/main/java/org/jboss/as/jpa/processor/JPAClassFileTransformerProcessor.java]: This should probably use `Boolean.parseBoolean(WildFlySecurityManager.getPropertyPrivileged("Hibernate51CompatibilityTransformer", "false))`.

File: testsuite/compat/src/test/java/org/jboss/as/test/compat/jpa/hibernate/transformer/VerifyHibernate51CompatibilityTestCase.java [hunk 3]
```diff
@@ -0,0 +1,404 @@
...
+    @Test
+    public void testSimpleOperation() throws Exception {
+        SFSBHibernateSessionFactory sfsb = lookup("SFSBHibernateSessionFactory", SFSBHibernateSessionFactory.class);
+        // setup Configuration and SessionFactory
+        sfsb.setupConfig();
+        try {
+            Student s1 = sfsb.createStudent("MADHUMITA", "SADHUKHAN", "99 Purkynova REDHAT BRNO CZ", 1);
+            Student s2 = sfsb.createStudent("REDHAT", "LINUX", "Worldwide", 3);
+            Student st = sfsb.getStudent(s1.getStudentId());
+            assertTrue( "name read from hibernate session is MADHUMITA", "MADHUMITA".equals( st.getFirstName() ) );
```
Comment (jamezp) [inline testsuite/compat/src/test/java/org/jboss/as/test/compat/jpa/hibernate/transformer/VerifyHibernate51CompatibilityTestCase.java]: It's not really a big deal, but is there a reason we're not using `assertEquals()` here?

File: jpa/subsystem/src/main/java/org/jboss/as/jpa/Hibernate51CompatibilityTransformer.java (hunk 1 above)
Comment (scottmarlow) [reply jpa/subsystem/src/main/java/org/jboss/as/jpa/Hibernate51CompatibilityTransformer.java]: hmm, I missed that, I actually think that I need asm-commons again (just removed it yesterday), for the next transformation needed.  I'll remove the todo for now.

Comment (scottmarlow) [reply jpa/subsystem/src/main/java/org/jboss/as/jpa/Hibernate51CompatibilityTransformer.java]: yes, thanks

File: jpa/subsystem/src/main/java/org/jboss/as/jpa/processor/JPAClassFileTransformerProcessor.java (hunk 2 above)
Comment (scottmarlow) [reply jpa/subsystem/src/main/java/org/jboss/as/jpa/processor/JPAClassFileTransformerProcessor.java]: fixing.

Comment (scottmarlow) [reply jpa/subsystem/src/main/java/org/jboss/as/jpa/processor/JPAClassFileTransformerProcessor.java]: fixing.

**Comment by kabir** [comment]
@mjurc @simkam I'll merge this on Monday unless I hear some objections in the morning

**Comment by kabir** [comment]
@scottmarlow @mjurc @simkam Adding the hold label since I think the work is now happening on a feature branch?

**Comment by kabir** [comment]
Closing until this is ready for merge

**Pull request closed without merging**